### PR TITLE
feat: dyanmic finetuning

### PR DESCRIPTION
**Pull request by ved1beta**
# Description

Adds Dynamic Fine-Tuning ([Wu et al., 2025](https://arxiv.org/abs/2508.05629)) as an opt-in SFT loss via use_dft: true.

DFT reweights each token's cross-entropy by its own detached probability:

L_DFT = -Σ_t  sg[π(y_t | y_<t)] · log π(y_t | y_<t)

Supersedes #3360, reworked to follow the EAFT (#3366) pattern.

## How has this been tested?
manual runs + tests 
## AI Usage Disclaimer
claude

## Summary by CodeRabbit

* **New Features**
  * Added Dynamic Fine-Tuning (DFT), which weights training loss using each token’s predicted probability.
  * Added a DFT training example for Google Gemma 3 1B Instruct.
  * Updated the performance table with DFT and EAFT options.

* **Bug Fixes**
  * Added validation preventing DFT and EAFT from being enabled together.

* **Documentation**
  * Documented compatibility requirements for DFT and EAFT, including their use of full logits.

### CODE REVIEW

File: examples/dft/dft-example.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
+base_model: google/gemma-3-1b-it
+
+model_type: Gemma3ForCausalLM
+cls_model_config: Gemma3TextConfig
```
Comment (NanoCode012) [inline examples/dft/dft-example.yml:4]: not needed
```suggestion
```
